flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java

```diff
@@ -779,6 +816,22 @@ public void testDuplicateJobSubmissionWithRunningJobId() throws Throwable {
         assertFalse(maybeDuplicate.get().isGloballyTerminated());
     }
 
+    @Test
+    public void testClusterShutsDownGracefullyOnJobDriverError() throws Exception {
+        final MiniClusterConfiguration cfg = new MiniClusterConfiguration.Builder().build();
+        try (final MiniCluster miniCluster = new ApplicationMiniCluster(cfg, getFailingProgram())) {
+            miniCluster.start();
+            final CompletableFuture<ApplicationStatus> shutdownFuture =
+                    Iterables.getOnlyElement(miniCluster.getDispatcherShutdownFutures());
+            try {
+                shutdownFuture.get();
+                fail("Exception expected.");
+            } catch (Exception e) {
+                assertThat(e, FlinkMatchers.containsCause(FailingJob.FailingJobException.class));
+            }
+        }
+    }
```

Comment:
This test doesn't seem to belong into the `ApplicationDispatcherBootstrapTest`. Maybe it would be better suited for `ApplicationDispatcherBootstrapITCase`.

Reply:
👍 Also I'll change the test to reflect described scenario (dispatcher loses leadership).